From b4b2918b2ac67ba6553d79dbf1366cfc9b669080 Mon Sep 17 00:00:00 2001
From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Wed, 17 Dec 2025 14:35:37 +1000
Subject: [PATCH] RHEL: map dials to rings on the Intuos Pro 3rd Gen devices

These devices have relative dials but our userspace stack
(mutter/gnome-shell and Wayland) do not have support for relative dials.
This means they can't be configured and any events cannot be
passed through to the client.

Work around this for this set of devices by mapping the relative dials
to a ring. This isn't a perfect emulation - the last value is simply
remembered but we only get so much data from a dial.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
---
 meson.build                                   |  1 +
 quirks/30-vendor-wacom.quirks                 | 42 +++++++++
 src/evdev-tablet-pad.c                        | 92 +++++++++++++++++--
 src/evdev-tablet-pad.h                        |  5 +
 src/quirks.c                                  |  3 +-
 src/quirks.h                                  |  1 +
 test/litest-device-wacom-intuos-pro-3rd-pad.c | 86 +++++++++++++++++
 test/litest.h                                 |  1 +
 test/test-pad.c                               | 72 +++++++++++++++
 9 files changed, 292 insertions(+), 11 deletions(-)
 create mode 100644 test/litest-device-wacom-intuos-pro-3rd-pad.c

diff --git a/meson.build b/meson.build
index 0db0d1da3aba..ac58c7a1d734 100644
--- a/meson.build
+++ b/meson.build
@@ -948,6 +948,7 @@ if get_option('tests')
 		'test/litest-device-wacom-cintiq-pro-16-pen.c',
 		'test/litest-device-wacom-ekr.c',
 		'test/litest-device-wacom-hid4800-pen.c',
+		'test/litest-device-wacom-intuos-pro-3rd-pad.c',
 		'test/litest-device-wacom-intuos3-pad.c',
 		'test/litest-device-wacom-intuos5-finger.c',
 		'test/litest-device-wacom-intuos5-pad.c',
diff --git a/quirks/30-vendor-wacom.quirks b/quirks/30-vendor-wacom.quirks
index a7898b65ca1c..38fbc65e08ba 100644
--- a/quirks/30-vendor-wacom.quirks
+++ b/quirks/30-vendor-wacom.quirks
@@ -40,3 +40,45 @@ MatchBus=usb
 MatchVendor=0x2D1F
 MatchProduct=0x524C
 AttrEventCode=-ABS_TILT_X;-ABS_TILT_Y;
+
+[Wacom Intuos Pro 3rd L (0x03F9)]
+MatchUdevType=tablet
+MatchVendor=0x056A
+MatchProduct=0x03F9
+ModelWacomIntuosPro3rd=1
+
+[Wacom Intuos Pro 3rd L (0x03FA)]
+MatchUdevType=tablet
+MatchVendor=0x056A
+MatchProduct=0x03FA
+ModelWacomIntuosPro3rd=1
+
+[Wacom Intuos Pro 3rd L (0x0401)]
+MatchUdevType=tablet
+MatchVendor=0x056A
+MatchProduct=0x0401
+ModelWacomIntuosPro3rd=1
+
+[Wacom Intuos Pro 3rd M (0x03F7)]
+MatchUdevType=tablet
+MatchVendor=0x056A
+MatchProduct=0x03F7
+ModelWacomIntuosPro3rd=1
+
+[Wacom Intuos Pro 3rd M (0x03F9)]
+MatchUdevType=tablet
+MatchVendor=0x056A
+MatchProduct=0x03F9
+ModelWacomIntuosPro3rd=1
+
+[Wacom Intuos Pro 3rd S (0x03F6)]
+MatchUdevType=tablet
+MatchVendor=0x056A
+MatchProduct=0x03F6
+ModelWacomIntuosPro3rd=1
+
+[Wacom Intuos Pro 3rd S (0x03F5)]
+MatchUdevType=tablet
+MatchVendor=0x056A
+MatchProduct=0x03F5
+ModelWacomIntuosPro3rd=1
diff --git a/src/evdev-tablet-pad.c b/src/evdev-tablet-pad.c
index 186def190a71..40552603f80d 100644
--- a/src/evdev-tablet-pad.c
+++ b/src/evdev-tablet-pad.c
@@ -354,6 +354,30 @@ pad_strip_get_mode_group(struct pad_dispatch *pad, unsigned int strip)
 	return NULL;
 }
 
+static double
+pad_ring_to_dial(double *current_value, double delta)
+{
+	/* The dial value is in v120 range but needs to be mapped to the 0..360
+	 * degrees that the ring provides.
+	 *
+	 * Let's say one wheel detent (v120) is 15 degrees, this gives us 24
+	 * lores wheel clicks to go the full 360 degrees circle.
+	 * If we have 24 clicks per 360 that means our max value is 120 * 24
+	 * after which we wrap around.
+	 */
+	const int dial_degrees = 15;
+	const int detents_per_360 = 360 / dial_degrees;
+	const int wrap_threshold = 120 * detents_per_360;
+
+	double abs_value = *current_value + delta;
+	abs_value = fmod(abs_value + wrap_threshold, wrap_threshold);
+	*current_value = abs_value;
+
+	double degrees = abs_value * dial_degrees / 120;
+
+	return degrees;
+}
+
 static void
 pad_check_notify_axes(struct pad_dispatch *pad,
 		      struct evdev_device *device,
@@ -372,18 +396,48 @@ pad_check_notify_axes(struct pad_dispatch *pad,
 
 	/* Unlike the ring axis we don't get an event when we release
 	 * so we can't set a source */
-	if (pad->changed_axes & PAD_AXIS_DIAL1) {
+	if (pad->changed_axes & PAD_AXIS_DIAL1 && !pad->dials.map_to_ring) {
 		group = pad_dial_get_mode_group(pad, 0);
 		if (group)
 			tablet_pad_notify_dial(base, time, 0, pad->dials.dial1, group);
 	}
 
-	if (pad->changed_axes & PAD_AXIS_DIAL2) {
+	if (pad->changed_axes & PAD_AXIS_DIAL2 && !pad->dials.map_to_ring) {
 		group = pad_dial_get_mode_group(pad, 1);
 		if (group)
 			tablet_pad_notify_dial(base, time, 1, pad->dials.dial2, group);
 	}
 
+	if (pad->changed_axes & PAD_AXIS_DIAL1 && pad->dials.map_to_ring) {
+		double degrees =
+			pad_ring_to_dial(&pad->dials.abs_dial1_value, pad->dials.dial1);
+		if (device->left_handed.enabled)
+			degrees = fmod(degrees + 180, 360);
+
+		group = pad_ring_get_mode_group(pad, 0);
+		tablet_pad_notify_ring(base,
+				       time,
+				       0,
+				       degrees,
+				       LIBINPUT_TABLET_PAD_RING_SOURCE_UNKNOWN,
+				       group);
+	}
+
+	if (pad->changed_axes & PAD_AXIS_DIAL2 && pad->dials.map_to_ring) {
+		double degrees =
+			pad_ring_to_dial(&pad->dials.abs_dial2_value, pad->dials.dial2);
+		if (device->left_handed.enabled)
+			degrees = fmod(degrees + 180, 360);
+
+		group = pad_ring_get_mode_group(pad, 1);
+		tablet_pad_notify_ring(base,
+				       time,
+				       1,
+				       degrees,
+				       LIBINPUT_TABLET_PAD_RING_SOURCE_UNKNOWN,
+				       group);
+	}
+
 	if (pad->changed_axes & PAD_AXIS_RING1) {
 		value = pad_handle_ring(pad, device, ABS_WHEEL);
 		if (send_finger_up)
@@ -824,6 +878,10 @@ pad_init(struct pad_dispatch *pad, struct evdev_device *device)
 	pad->changed_axes = PAD_AXIS_NONE;
 	ratelimit_init(&pad->modes.group_not_found, h2us(1), 3);
 
+	/* RHEL dial-to-ring special */
+	pad->dials.map_to_ring =
+		evdev_device_has_model_quirk(device, QUIRK_MODEL_WACOM_INTUOS_PRO_3RD);
+
 	/* We expect the kernel to either give us both axes as hires or neither.
 	 * Getting one is a kernel bug we don't need to care about */
 	pad->dials.has_hires_dial =
@@ -890,24 +948,34 @@ evdev_device_tablet_pad_get_num_buttons(struct evdev_device *device)
 	return pad->nbuttons;
 }
 
-int
-evdev_device_tablet_pad_get_num_dials(struct evdev_device *device)
+static int
+count_dials(struct libevdev *evdev)
 {
 	int ndials = 0;
 
-	if (!(device->seat_caps & EVDEV_DEVICE_TABLET_PAD))
-		return -1;
-
-	if (libevdev_has_event_code(device->evdev, EV_REL, REL_WHEEL) ||
-	    libevdev_has_event_code(device->evdev, EV_REL, REL_DIAL)) {
+	if (libevdev_has_event_code(evdev, EV_REL, REL_WHEEL) ||
+	    libevdev_has_event_code(evdev, EV_REL, REL_DIAL)) {
 		ndials++;
-		if (libevdev_has_event_code(device->evdev, EV_REL, REL_HWHEEL))
+		if (libevdev_has_event_code(evdev, EV_REL, REL_HWHEEL))
 			ndials++;
 	}
 
 	return ndials;
 }
 
+int
+evdev_device_tablet_pad_get_num_dials(struct evdev_device *device)
+{
+	if (!(device->seat_caps & EVDEV_DEVICE_TABLET_PAD))
+		return -1;
+
+	/* RHEL dial-to-ring special */
+	if (evdev_device_has_model_quirk(device, QUIRK_MODEL_WACOM_INTUOS_PRO_3RD))
+		return 0;
+
+	return count_dials(device->evdev);
+}
+
 int
 evdev_device_tablet_pad_get_num_rings(struct evdev_device *device)
 {
@@ -916,6 +984,10 @@ evdev_device_tablet_pad_get_num_rings(struct evdev_device *device)
 	if (!(device->seat_caps & EVDEV_DEVICE_TABLET_PAD))
 		return -1;
 
+	/* RHEL dial-to-ring special */
+	if (evdev_device_has_model_quirk(device, QUIRK_MODEL_WACOM_INTUOS_PRO_3RD))
+		return count_dials(device->evdev);
+
 	if (libevdev_has_event_code(device->evdev, EV_ABS, ABS_WHEEL)) {
 		nrings++;
 		if (libevdev_has_event_code(device->evdev, EV_ABS, ABS_THROTTLE))
diff --git a/src/evdev-tablet-pad.h b/src/evdev-tablet-pad.h
index 27e9fbed47c3..4c9fef3200f5 100644
--- a/src/evdev-tablet-pad.h
+++ b/src/evdev-tablet-pad.h
@@ -83,6 +83,11 @@ struct pad_dispatch {
 		bool has_hires_dial;
 		double dial1;
 		double dial2;
+
+		/* RHEL: map the dial to ring on selected devices */
+		bool map_to_ring;
+		double abs_dial1_value;
+		double abs_dial2_value;
 	} dials;
 
 	struct {
diff --git a/src/quirks.c b/src/quirks.c
index 6364549c8708..c726e03eb131 100644
--- a/src/quirks.c
+++ b/src/quirks.c
@@ -298,7 +298,8 @@ quirk_get_name(enum quirk q)
 		return "ModelWacomTouchpad";
 	case QUIRK_MODEL_PRESSURE_PAD:
 		return "ModelPressurePad";
-
+	case QUIRK_MODEL_WACOM_INTUOS_PRO_3RD:
+		return "ModelWacomIntuosPro3rd";
 	case QUIRK_ATTR_SIZE_HINT:
 		return "AttrSizeHint";
 	case QUIRK_ATTR_TOUCH_SIZE_RANGE:
diff --git a/src/quirks.h b/src/quirks.h
index b6ea220a991c..22321d644f12 100644
--- a/src/quirks.h
+++ b/src/quirks.h
@@ -95,6 +95,7 @@ enum quirk {
 	QUIRK_MODEL_TOUCHPAD_VISIBLE_MARKER,
 	QUIRK_MODEL_TRACKBALL,
 	QUIRK_MODEL_WACOM_TOUCHPAD,
+	QUIRK_MODEL_WACOM_INTUOS_PRO_3RD,
 
 	_QUIRK_LAST_MODEL_QUIRK_, /* Guard: do not modify */
 
diff --git a/test/litest-device-wacom-intuos-pro-3rd-pad.c b/test/litest-device-wacom-intuos-pro-3rd-pad.c
new file mode 100644
index 000000000000..8efdcb16740b
--- /dev/null
+++ b/test/litest-device-wacom-intuos-pro-3rd-pad.c
@@ -0,0 +1,86 @@
+/*
+ * Copyright © 2016 Red Hat, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "config.h"
+
+#include "litest-int.h"
+#include "litest.h"
+
+static struct input_event down[] = {
+	{ .type = -1, .code = -1 },
+};
+
+static struct input_event move[] = {
+	{ .type = -1, .code = -1 },
+};
+
+static struct litest_device_interface interface = {
+	.touch_down_events = down,
+	.touch_move_events = move,
+};
+
+static struct input_absinfo absinfo[] = {
+	{ ABS_X, 0, 1, 0, 0, 0 },
+	{ ABS_Y, 0, 1, 0, 0, 0 },
+	{ ABS_MISC, 0, 0, 0, 0, 0 },
+	{ .value = -1 },
+};
+
+static struct input_id input_id = {
+	.bustype = 0x3,
+	.vendor = 0x56a,
+	.product = 0x3f9,
+};
+
+static int events[] = {
+	EV_KEY, BTN_0,
+	EV_KEY, BTN_1,
+	EV_KEY, BTN_2,
+	EV_KEY, BTN_3,
+	EV_KEY, BTN_4,
+	EV_KEY, BTN_5,
+	EV_KEY, BTN_6,
+	EV_KEY, BTN_7,
+	EV_KEY, BTN_8,
+	EV_KEY, BTN_9,
+	EV_KEY, BTN_STYLUS,
+	EV_REL, REL_WHEEL,
+	EV_REL, REL_HWHEEL,
+	EV_REL, REL_WHEEL_HI_RES,
+	EV_REL, REL_HWHEEL_HI_RES,
+	-1,     -1,
+};
+
+TEST_DEVICE(
+	LITEST_WACOM_INTUOS_PRO_3RD_PAD,
+	.features = LITEST_TABLET_PAD, /* RHEL: not LITEST_DIAL because it's special */
+	.interface = &interface,
+
+	.name = "Wacom Intuos Pro L Pad",
+	.id = &input_id,
+	.events = events,
+	.absinfo = absinfo,
+	.udev_properties = {
+		{ "ID_INPUT_TABLET_PAD", "1" },
+		{ NULL },
+	}, )
diff --git a/test/litest.h b/test/litest.h
index 9aa76f3c27ca..fe88be22e984 100644
--- a/test/litest.h
+++ b/test/litest.h
@@ -532,6 +532,7 @@ enum litest_device_type {
 	LITEST_WACOM_CINTIQ_PRO16_PEN,
 	LITEST_WACOM_EKR,
 	LITEST_WACOM_HID4800_PEN,
+	LITEST_WACOM_INTUOS_PRO_3RD_PAD,
 	LITEST_WACOM_INTUOS3_PAD,
 	LITEST_WACOM_INTUOS5_PAD,
 	LITEST_WACOM_INTUOS5_PEN,
diff --git a/test/test-pad.c b/test/test-pad.c
index 6c3560d6bd38..2bac2f7866f1 100644
--- a/test/test-pad.c
+++ b/test/test-pad.c
@@ -1147,6 +1147,67 @@ START_TEST(pad_send_events_disabled)
 }
 END_TEST
 
+START_TEST(pad_dial_to_ring)
+{
+	struct litest_device *dev = litest_current_device();
+	struct libinput *li = dev->libinput;
+	struct libinput_event *event;
+	struct libinput_event_tablet_pad *pev;
+
+	unsigned int ring = litest_test_param_get_u32(test_env->params, "ring");
+	int direction = litest_test_param_get_i32(test_env->params, "direction");
+	bool left_handed = litest_test_param_get_bool(test_env->params, "left-handed");
+
+	const int degrees_per_detent = 15;
+
+	litest_assert_int_eq(
+		libinput_device_tablet_pad_get_num_dials(dev->libinput_device),
+		0);
+	litest_assert_int_eq(
+		libinput_device_tablet_pad_get_num_rings(dev->libinput_device),
+		2);
+
+	/* The dial-to-ring affected devices do not support left-handed so let's
+	 * make sure setting left-handed a) doesn't work and b) the setting
+	 * doesn't affect anything.
+	 */
+	enum libinput_config_status rc =
+		libinput_device_config_left_handed_set(dev->libinput_device,
+						       left_handed);
+	litest_assert_int_eq(rc, (unsigned)LIBINPUT_CONFIG_STATUS_UNSUPPORTED);
+
+	litest_drain_events(li);
+
+	double expected = 0.0;
+	for (size_t count = 0; count < 36;
+	     count++) { /* 24 per 360 so let's do 1.5 rotations */
+		if (ring == 0) {
+			/* REL_WHEEL is inverted to expectations */
+			litest_event(dev, EV_REL, REL_WHEEL, 1 * -direction);
+			litest_event(dev, EV_REL, REL_WHEEL_HI_RES, 120 * -direction);
+			litest_event(dev, EV_SYN, SYN_REPORT, 0);
+		} else {
+			litest_event(dev, EV_REL, REL_HWHEEL, 1 * direction);
+			litest_event(dev, EV_REL, REL_HWHEEL_HI_RES, 120 * direction);
+			litest_event(dev, EV_SYN, SYN_REPORT, 0);
+		}
+
+		libinput_dispatch(li);
+
+		expected += 360 + direction * degrees_per_detent;
+		expected = fmod(expected, 360);
+
+		event = libinput_get_event(li);
+		pev = litest_is_pad_ring_event(event,
+					       ring,
+					       LIBINPUT_TABLET_PAD_RING_SOURCE_UNKNOWN);
+		double value = libinput_event_tablet_pad_get_ring_position(pev);
+		litest_assert_double_eq(value, expected);
+		libinput_event_destroy(event);
+	}
+}
+END_TEST
+
 TEST_COLLECTION(pad)
 {
 	/* clang-format off */
@@ -1192,5 +1253,16 @@ TEST_COLLECTION(pad)
 	litest_add(pad_keys, LITEST_TABLET_PAD, LITEST_ANY);
 
 	litest_add(pad_send_events_disabled, LITEST_TABLET_PAD, LITEST_ANY);
+
+	enum {
+		CCW = -1,
+		CW = 1,
+	};
+	litest_with_parameters(params,
+			       "ring", 'u', 2, 0, 1,
+			       "direction", 'I', 2, litest_named_i32(CCW), litest_named_i32(CW),
+			       "left-handed", 'b') {
+		litest_add_parametrized_for_device(pad_dial_to_ring, LITEST_WACOM_INTUOS_PRO_3RD_PAD, params);
+	}
 	/* clang-format on */
 }
-- 
2.51.1

